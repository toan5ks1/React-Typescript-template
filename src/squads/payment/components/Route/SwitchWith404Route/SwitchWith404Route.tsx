import { ReactNode } from "react";

import { Redirect, Route, Switch, SwitchProps, useLocation } from "react-router";

interface SwitchWith404RouteProps extends SwitchProps {
    catchAllMode?: "404" | "component";
    component?: ReactNode;
    basename?: string;
}
const SwitchWith404Route = (props: SwitchWith404RouteProps) => {
    const { children, catchAllMode = "404", component } = props;
    const { pathname } = useLocation();
    const shouldRedirect = pathname !== "/";

    return (
        <Switch>
            {children}
            <Route
                render={() => {
                    if (catchAllMode === "404") {
                        if (shouldRedirect) {
                            return <Redirect to="/page-not-found" />;
                        }
                    }

                    if (catchAllMode === "component") return component;
                }}
            />
        </Switch>
    );
};

export default SwitchWith404Route;
