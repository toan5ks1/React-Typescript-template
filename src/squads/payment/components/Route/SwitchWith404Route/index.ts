export { default } from "./SwitchWith404Route";
