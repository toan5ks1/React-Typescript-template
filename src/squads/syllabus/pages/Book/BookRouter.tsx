import { useRouteMatch } from "react-router";
import { lazyWithRetry } from "src/common/utils/other";

import RouteGuard from "src/components/Route/RouteGuard";
import SwitchWith404Route from "src/components/Route/SwitchWith404Route";

const BookDetailsPageLazy = lazyWithRetry(() => import("./BookDetails"));
const BookListPageLazy = lazyWithRetry(() => import("./BookList"));

const BookRouter = () => {
    const { path } = useRouteMatch();
    return (
        <SwitchWith404Route>
            <RouteGuard path={`${path}/:id/show`} component={BookDetailsPageLazy} />
            <RouteGuard path={path} component={BookListPageLazy} />
        </SwitchWith404Route>
    );
};

export default BookRouter;
