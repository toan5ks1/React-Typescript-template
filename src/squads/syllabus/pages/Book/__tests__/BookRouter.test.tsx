import { Suspense } from "react";

import { MemoryRouter, Route } from "react-router";

import SwitchWith404Route from "src/components/Route/SwitchWith404Route";

import BookRouter from "../BookRouter";

import { render, waitFor, screen } from "@testing-library/react";

jest.mock("src/squads/syllabus/pages/Book/BookList", () => {
    return {
        __esModule: true,
        default: () => {
            return <div>BookList</div>;
        },
    };
});

describe("<BookRouter />", () => {
    it("should render book list without crash", async () => {
        const { container } = render(
            <MemoryRouter initialEntries={["/syllabus/books"]}>
                <Suspense fallback={"..."}>
                    <SwitchWith404Route>
                        <Route path="/syllabus/books" component={BookRouter} />
                    </SwitchWith404Route>
                </Suspense>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.queryByText("...")).not.toBeInTheDocument();
        });

        expect(container).toMatchSnapshot();
    });
});
