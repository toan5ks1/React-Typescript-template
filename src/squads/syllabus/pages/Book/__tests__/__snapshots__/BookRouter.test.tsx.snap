// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookRouter /> should render book list without crash 1`] = `
<div>
  <div>
    BookList
  </div>
</div>
`;
